google/cloud/internal/rest_completion_queue_impl.cc

```diff
@@ -0,0 +1,68 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/rest_completion_queue_impl.h"
+#include "google/cloud/internal/timer_queue.h"
+
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+RestCompletionQueueImpl::RestCompletionQueueImpl()
+    : tq_(std::make_shared<internal::TimerQueue>()) {}
+
+void RestCompletionQueueImpl::Run() { tq_->Service(); }
+
+void RestCompletionQueueImpl::Shutdown() { tq_->Shutdown(); }
+
+void RestCompletionQueueImpl::CancelAll() { tq_->CancelAll(); }
+
+future<StatusOr<std::chrono::system_clock::time_point>>
+RestCompletionQueueImpl::MakeDeadlineTimer(
+    std::chrono::system_clock::time_point deadline) {
+  return tq_->Schedule(deadline);
+}
+
+future<StatusOr<std::chrono::system_clock::time_point>>
+RestCompletionQueueImpl::MakeRelativeTimer(std::chrono::nanoseconds duration) {
+  using std::chrono::system_clock;
+  auto const d = std::chrono::duration_cast<system_clock::duration>(duration);
+  return MakeDeadlineTimer(system_clock::now() + d);
+}
+
+// Use an "immediately" expiring timer in order to get the thread(s) servicing
+// the TimerQueue to execute the function. However, if the timer
+// expires before .then() is invoked, the lambda will be immediately called and
+// the passing of execution to the queue servicing thread will not occur.
+void RestCompletionQueueImpl::RunAsync(
+    std::unique_ptr<internal::RunAsyncBase> function) {
+  ++run_async_counter_;
+  // Some delay is necessary to reduce the likelihood that the timer expires
+  // before .then() is called.
```

Comment:
Aside:  I'm afraid I don't grok that.  It is the timer expiring that invokes the `.then()`, no?

Reply:
Consider:

```
promise<int> p;
auto f = p.get_future();
```

In this code the assertion will succeed, because the future was already satisfied when you called `.then()` and therefore the lambda runs immediately upon calling `.then()`:

```
p.set_value(42);
auto g = f.then([](auto) { return std::this_thread::get_id(); });
assert(g.get() == std::this_thread::get_id());
```

In this code the assertion will fail, because the future is not satisfied when `.then()` is called and therefore the lambda runs in the thread that satisfies the future:

```
auto g = f.then([](auto) { return std::this_thread::get_id(); });
auto t = std::thread([](promise<int> p) { p.set_value(42); }, std::move(p); );
assert(g.get() == std::this_thread::get_id());
```

In this code it is unknown if the assertion will succeed or fail because the future may or may not be called by the time `.then()` is called.

```
auto t = std::thread([](promise<int> p) { p.set_value(42); }, std::move(p); );
auto g = f.then([](auto) { return std::this_thread::get_id(); });
assert(g.get() == std::this_thread::get_id());
t.join();
```

We need some changes in TimerQueue (#10553) to make sure `RunAsync()` behaves like the second case (except when timer queues are shutdown, but that is another story).

Reply:
>> Aside: I'm afraid I don't grok that. It is the timer expiring that invokes the .then(), no?
> 
> the future may or may not be [satisfied] by the time .then() is called.

OK.  Thanks.  Indeed, no.

I guess, then, the problem might be more clearly described as being about ensuring `exec()` is called from a service thread.

> We need some changes in TimerQueue (https://github.com/googleapis/google-cloud-cpp/issues/10553)

Ack.  The simplest approach (methinks) is unrelated to timers, so the young class's name has already started to lose some meaning.